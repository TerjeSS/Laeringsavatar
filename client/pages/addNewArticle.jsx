import React, { useContext, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { FormInput } from "../lib/formInput";
import { FormTextArea } from "../lib/formTextArea";
import { ArticlesApiContext } from "../articlesApiContext";

export function AddNewArticle() {
  const { createArticle } = useContext(ArticlesApiContext);
  const [title, setTitle] = useState("");
  const [date, setDate] = useState("");
  const [author, setAuthor] = useState("");
  const [topic, setTopic] = useState("");
  const [articletext, setArticletext] = useState("");

  const navigate = useNavigate();

  function handleSubmit(event) {
    event.preventDefault();
    createArticle({ title, date: parseInt(date), author, topic, articletext });
    navigate("/articles");
  }

  return (
    <form onSubmit={handleSubmit}>
      <div id={"back_link"}>
        <Link to={"/profile"}>Tilbake</Link>
      </div>
      <h1>Add new article</h1>
      <FormInput label={"Title:"} value={title} onChangeValue={setTitle} />
      <FormInput label={"Year:"} value={date} onChangeValue={setDate} />
      <FormInput label={"Author:"} value={author} onChangeValue={setAuthor} />
      <FormInput label={"Topic:"} value={topic} onChangeValue={setTopic} />
      <br />
      <FormTextArea
        label={"Article text:"}
        value={articletext}
        setValue={setArticletext}
      />
      <br />

      <p>Title, topic and article text must be filled out!</p>
      <div>
        <button
          disabled={
            title.length === 0 || topic.length === 0 || articletext.length === 0
          }
        >
          Save
        </button>
      </div>
    </form>
  );
}
