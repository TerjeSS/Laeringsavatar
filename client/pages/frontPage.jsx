import React, { useContext } from "react";
import { ArticlesApiContext } from "../articlesApiContext";
import { useLoading } from "../useLoading";
import { Link } from "react-router-dom";

export function FrontPage() {
  const { listArticles } = useContext(ArticlesApiContext);
  const { loading, error, data } = useLoading(listArticles);

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>VisMo3D</h1>
      <p>
        VisMo3D er lorem ipsum dolor sit amet, consectetur adipiscing elit, sed
        do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
        ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
        aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit
        in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
        Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
        officia deserunt mollit anim id est laborum.
      </p>
      <Link to={"/login"}>
        <h3>Logg inn</h3>
      </Link>
    </div>
  );
}
