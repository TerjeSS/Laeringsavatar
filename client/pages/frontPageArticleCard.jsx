import React from "react";

export function FrontPageArticleCard({ article: { title, topic, poster } }) {
  return (
    <>
      <div id={"article_card_info"}>
        <h1>{title}</h1>
        <h3>{topic}</h3>
        <img src={poster} alt="Fil" width={300} height={200} />
      </div>
    </>
  );
}
