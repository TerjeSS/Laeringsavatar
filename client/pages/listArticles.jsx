import React, { useContext } from "react";
import { useLoading } from "../useLoading";
import { ArticlesApiContext } from "../articlesApiContext";

function ArticleCard({
  article: { title, date, author, poster, topic, articletext },
}) {
  return (
    <>
      <div id={"article_card_info"}>
        <h1>Navn: {title}</h1>
        <h3>Skole: {topic}</h3>
        <h3>Sted: {author}</h3>
        <img src={poster} alt="Fil" width={300} height={200} />
      </div>
    </>
  );
}

export function ListArticles() {
  const { listArticles } = useContext(ArticlesApiContext);
  const { loading, error, data } = useLoading(listArticles);

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      {data.map((article) => (
        <ArticleCard key={article.topic} article={article} />
      ))}
    </div>
  );
}
