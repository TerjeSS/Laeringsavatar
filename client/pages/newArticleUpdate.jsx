import React, { useEffect, useState } from "react";

function ArticleUpdate({ news: { headline, message } }) {
  return (
    <div>
      <strong>{headline}: </strong>
      {message}
    </div>
  );
}

function ArticleUpdateApplication() {
  const [ws, setWs] = useState();
  useEffect(() => {
    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    ws.onmessage = (event) => {
      const { headline, message } = JSON.parse(event.data);
      setNewsLog((oldState) => [...oldState, { headline, message }]);
    };
    setWs(ws);
  }, []);

  const [newsLog, setNewsLog] = useState([]);
  return (
    <div>
      <h1>Live nyheter</h1>
      <main>
        {newsLog.map((news, index) => (
          <ArticleUpdate key={index} news={news} />
        ))}
      </main>
      <h1>Siden er under p√•bygging!</h1>
    </div>
  );
}

export function NewArticleUpdate() {
  return <ArticleUpdateApplication />;
}
