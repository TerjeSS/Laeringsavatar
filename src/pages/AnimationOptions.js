import GUI from 'lil-gui'; 
import {SkinHandling} from "./SkinHandling.js"

let THREE;

export class SkinOptions {
    constructor(three, skin, pants, glasses, mixer, div) {
      THREE = three;
      this.skin = skin;
      this.pants = pants;
      this.glasses = glasses;
      this.mixer = mixer;
      this.skinHandler = new SkinHandling(THREE);

      // Create a new GUI instance
      this.panel = new GUI( {title: "⚙️", container: div} );
      this.speed = this.panel.addFolder("Avspilling");
      this.hud = this.panel.addFolder('Hud');
      this.bukser = this.panel.addFolder('Bukser');
      this.glass = this.panel.addFolder('Briller');

      var settings = {
        'Treningsbukse': () => this.SetTrackPants(),
        'Joggebukse': () => this.SetCottonPants(),
        'Jeans': () => this.SetJeans(),
        'Lærbukse': () => this.SetLeatherPants(),
        'Slangebukse': () => this.SetSnakePants(),
        'Gullbukse': () => this.SetGoldPants(),
        'Mesh': () => this.SetMesh(),
        'Blank': () => this.SetBlank(),
        'Slange': () => this.SetSnakeSkin(),
        'Gull': () => this.SetGoldSkin(),
        'Plast': () => this.SetPlasticSkin()
      };

      this.mesh = this.hud.add(settings, "Mesh");
      this.mesh.disable();
      this.blank = this.hud.add(settings, 'Blank');
      this.snakeSkin = this.hud.add(settings, 'Slange');
      this.goldSkin = this.hud.add(settings, 'Gull');
      this.plasticSkin = this.hud.add(settings, 'Plast');
      this.hud.addColor(skin.material, 'color').name( 'Farge' );

      this.trackPants = this.bukser.add(settings, 'Treningsbukse');
      this.trackPants.disable();
      this.cottonPants = this.bukser.add(settings, 'Joggebukse');
      this.jeans = this.bukser.add(settings, 'Jeans'); 
      this.leatherPants = this.bukser.add(settings, 'Lærbukse');     
      this.snakePants = this.bukser.add(settings, 'Slangebukse');
      this.goldPants = this.bukser.add(settings, 'Gullbukse');
      this.bukser.addColor(pants.material, 'color').name( 'Farge' );

      this.glass.add({'Briller': true}, 'Briller').onChange( () => this.GlassVisibility() );
      this.glass.addColor( this.glasses.material, 'color' ).name( 'Farge' ).onChange( () => {
        this.skinHandler.removeGlassMaps(this.glasses.material);
      });
      this.glass.add( { reset: () => {
        if (!this.glasses.material.map) {
            this.skinHandler.loadGlasses(this.glasses.material);
        }
       }}, 'reset' ).name( 'Svarte briller' );

      this.speed.add({'Hastighet':1.0}, 'Hastighet', 0.0,2.0,0.01).onChange((speed) => {
        this.mixer.timeScale = speed;
      });
      
      this.panel.close();
    }

    EnableAllSkins()
    {
        this.mesh.enable();
        this.blank.enable();
        this.snakeSkin.enable();
        this.goldSkin.enable();
        this.plasticSkin.enable();
        this.skin.material.transparent = false;
    }

    SetMesh()
    {
        this.skinHandler.loadDefaultManequin(this.skin.material);
        this.EnableAllSkins();
        this.mesh.disable();
    }

    SetBlank()
    {
        this.skinHandler.loadGreyManequin(this.skin.material);
        this.EnableAllSkins();
        this.blank.disable();
    }

    SetSnakeSkin()
    {
        this.skinHandler.loadSnakeSkin(this.skin.material);
        this.EnableAllSkins();
        this.snakeSkin.disable();
    }

    SetGoldSkin()
    {
        this.skinHandler.loadGold(this.skin.material);
        this.EnableAllSkins();
        this.goldSkin.disable();
    }

    SetPlasticSkin()
    {
        this.skinHandler.loadPolypropylene(this.skin.material);
        this.EnableAllSkins();
        this.skin.material.transparent = true;
        this.skin.material.opacity = 0.75;
        this.plasticSkin.disable();
    }

    SetTrackPants()
    {
        this.skinHandler.loadDefaultPants(this.pants.material);
        this.EnableAllPants();
        this.trackPants.disable();
    }

    SetSnakePants()
    {
        this.skinHandler.loadSnakeSkin(this.pants.material)
        this.EnableAllPants();
        this.snakePants.disable();
    }

    SetCottonPants()
    {
        this.skinHandler.loadCotton(this.pants.material);
        this.EnableAllPants();
        this.cottonPants.disable();
    }

    SetJeans()
    {
        this.skinHandler.loadDenim(this.pants.material);
        this.EnableAllPants();
        this.jeans.disable();
    }

    SetGoldPants()
    {
        this.skinHandler.loadGold(this.pants.material);
        this.EnableAllPants();
        this.goldPants.disable();
    }

    SetLeatherPants()
    {
        this.skinHandler.loadLeather(this.pants.material);
        this.EnableAllPants();
        this.leatherPants.disable();
    }

    EnableAllPants(){
        this.trackPants.enable();
        this.snakePants.enable();
        this.cottonPants.enable();
        this.jeans.enable();
        this.goldPants.enable();
        this.leatherPants.enable();
    }
  
    GlassVisibility()
    {
        this.glasses.visible = !this.glasses.visible;
    }
  
  }