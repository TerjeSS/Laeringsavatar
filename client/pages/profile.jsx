import { useLoading } from "../useLoading";
import { Link } from "react-router-dom";
import { FrontPageArticleCard } from "./frontPageArticleCard";
import React, { useContext } from "react";
import { ArticlesApiContext } from "../articlesApiContext";
import { FormInput } from "../lib/formInput";
import { FormTextArea } from "../lib/formTextArea";

import React, { useContext, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { FormInput } from "../lib/formInput";
import { FormTextArea } from "../lib/formTextArea";
import { ArticlesApiContext } from "../articlesApiContext";

export function Profile({ user }) {
  const { listArticles } = useContext(ArticlesApiContext);
  const { loading, error, data } = useLoading(listArticles);
  const { createArticle } = useContext(ArticlesApiContext);
  const [title, setTitle] = useState("");
  const [date, setDate] = useState("");
  const [author, setAuthor] = useState("");
  const [topic, setTopic] = useState("");
  const [articletext, setArticletext] = useState("");

  const navigate = useNavigate();

  function handleSubmit(event) {
    event.preventDefault();
    createArticle({ title, date: parseInt(date), author, topic, articletext });
    navigate("/articles");
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }
  return (
    <div>
      <div id={"userinfo"}>
        <div>
          <h1 className={"googleName"}>{user.google.name}</h1>
        </div>
        <div>
          <h1 className={"googleEmail"}>{user.google.email}</h1>
        </div>
        <div>
          <img src={user.google.picture} alt={"Profile picture"} />
        </div>
      </div>
      <form onSubmit={handleSubmit}>
        <h1>Legg til </h1>
        <FormInput label={"Navn:"} value={title} onChangeValue={setTitle} />
        <FormInput label={""} value={date} onChangeValue={setDate} />
        <FormInput label={"Sted:"} value={author} onChangeValue={setAuthor} />
        <FormInput label={"Skole:"} value={topic} onChangeValue={setTopic} />
        <br />
        <FormTextArea
          label={"Legg til fil:"}
          value={articletext}
          setValue={setArticletext}
        />
        <br />

        <p>Alle felter må være fylt ut for å publisere!</p>
        <div>
          <button
            disabled={
              title.length === 0 ||
              topic.length === 0 ||
              articletext.length === 0
            }
          >
            Publiser
          </button>
          <br /> <Link to={"/articles"}>Filer</Link>
        </div>
      </form>
    </div>
  );
}
