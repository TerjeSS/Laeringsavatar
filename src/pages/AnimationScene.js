import React, { useEffect, useState } from "react";
import * as THREE from "three";
import { OrbitControls } from "three/addons/controls/OrbitControls.js";

import { GLTFLoader } from "three/addons/loaders/GLTFLoader.js";
import { DRACOLoader } from "three/addons/loaders/DRACOLoader.js";
import { getDownloadURL, listAll, ref } from "firebase/storage";

import ControlPanel from "../components/ControlPanel/ControlPanel";
import { storage } from "../resources/firebase";
import { useParams } from "react-router-dom";
import { SkinOptions } from "./AnimationOptions.js"

const AnimationScene = () => {
  const [firebaseURL, setFirebaseURL] = useState("");
  let fileReferences = [];
  const storageRef = ref(storage);
  const animationRef = ref(storageRef, "animations");
  const { filename } = useParams();
  let camera, sceneBB, controls, bones, centreBone, bonePos, targetGeometry;
  let renderer, scene, model, glasses, pants, mannequin, loader, mixer, clock;
  let selectingCenter = false;
  let singleStepMode = false;
  let stepSize = 0;

  const fetchRefereces = async () => {
    listAll(animationRef)
      .then((res) => {
        fileReferences = [...res.items];
        const referenceName = fileReferences.filter((element) => {
          return element.name === filename;
        });
        getDownloadURL(ref(storage, referenceName)).then((res) => {
          setFirebaseURL(res);
        });
      })
      .catch((error) => {
        console.log(error);
      });
  };

  useEffect(() => {
    fetchRefereces();
  }, []);

  setTimeout(() => {
    createScene();
  }, 1);

  //THREE.JS SECTION ***************
  function findChild(parent, childName) {
    for (var child of parent.children) {
      if (child.name == childName)
        return child;
    }
    return null;
  }
  function loadModel() {
    const dracoLoader = new DRACOLoader();
    dracoLoader.setDecoderPath("js/libs/draco/gltf/");
    const loader = new GLTFLoader();
    loader.setDRACOLoader(dracoLoader);
    loader.load(
      firebaseURL,
      function (gltf) {
        model = gltf.scene;
  //      model.visible = false;
        var skeleton = findChild(model, "GLB_GRP");
        if(!skeleton)
          skeleton = findChild(model, "mannequin_skeleton");
        glasses = findChild(skeleton, "Glasses_Square2");
        if(!glasses)
          glasses = findChild(skeleton, "Glasses_Square");
        pants = findChild(skeleton, "TrackFBXASC032pants2");
        if(!pants)
          pants = findChild(skeleton, "Track_pants");
        mannequin = findChild(skeleton, "mannequin2");
        if(!mannequin)
          mannequin = findChild(skeleton, "mannequin");
        bones = findChild(skeleton, "Hips");
        if(!bones)
        {
          var armature = findChild(skeleton, "Armature");
          bones = findChild(armature, "Hips_1");
        }
        centreBone = bones;
        scene.add(model);

        model.traverse(function (object) {
          if (object.isMesh) {
            object.castShadow = true;
            // increase bounding box to make raycasting work
            object.geometry.boundingBox = sceneBB.clone();
            object.geometry.boundingBox.expandByScalar(2);
            const center = new THREE.Vector3();
            sceneBB.getCenter(center);
            const radius = sceneBB.getSize(new THREE.Vector3()).length() / 2;
            object.geometry.boundingSphere = new THREE.Sphere(center, radius);
            object.geometry.boundingSphere.expandByPoint(new THREE.Vector3(7, 7, 7));
          }
        });

        mixer = new THREE.AnimationMixer(model);

        //Starts the animation
        var animationSequence = gltf.animations.slice(-1)[0];
        mixer.clipAction(animationSequence).play();

        //Adding event listeners for the buttons
        var toggleButton = document.querySelector("#toggle-button");
        toggleButton.addEventListener("change", function () {
          if (singleStepMode) {
            mixer.clipAction(animationSequence).paused = false;
            singleStepMode = false;
          }
          else
            mixer.clipAction(animationSequence).paused = !mixer.clipAction(animationSequence).paused;
          selectingCenter = !selectingCenter;

        });
        document
          .querySelector(".reset-button")
          .addEventListener("click", () => {
            mixer.clipAction(animationSequence).reset();
            selectingCenter = false;
            singleStepMode = false;
            toggleButton.checked = false;
          });
        document.querySelector(".next-button")
          .addEventListener("click", () => {
            mixer.clipAction(animationSequence).paused = false;
            toggleButton.checked = true;
            singleStepMode = true;
            stepSize = 0.05;
          });
        document.querySelector(".prev-button")
          .addEventListener("click", () => {
            mixer.clipAction(animationSequence).paused = false;
            toggleButton.checked = true;
            singleStepMode = true;
            stepSize = -0.05;
          });

        // remove loading screen and enable options
        const overlayDiv = document.querySelector('.loadingDiv');
        overlayDiv.style.display = 'none';
        const canvasDiv = document.querySelector(".optionDiv")
        const audioDiv = document.querySelector("audio");
        new SkinOptions(THREE, mannequin, pants, glasses, mixer, canvasDiv, audioDiv);
        animate();
      },
      undefined,
      function (e) {
        console.error(e);
      }
    );

  }

  function loadScene(scene) {
    loader = new GLTFLoader();
    loader.load("/Stage.glb", (gltf) => {
      gltf.scene.traverse(function (node) {
        if (node.isMesh) {
          node.receiveShadow = true;
        }
      });

      sceneBB = new THREE.Box3().setFromObject(gltf.scene);
      sceneBB.expandByScalar(3.0);
      scene.add(gltf.scene);

      loadModel();
    });
  }

  function createScene() {
    bonePos = new THREE.Vector3();

    clock = new THREE.Clock();
    const container = document.querySelector(".canvas");
    window.addEventListener('resize', resize, false);

    renderer = new THREE.WebGLRenderer({
      canvas: container,
      antialias: true,
    });
    renderer.setPixelRatio(window.devicePixelRatio);
    renderer.setSize(window.innerWidth, window.innerHeight);
    renderer.outputColorSpace = THREE.SRGBColorSpace;
    renderer.shadowMap.enabled = true;
    renderer.useLegacyLights = true;

    scene = new THREE.Scene();
    scene.background = new THREE.Color(0x000000);

    //Camera
    camera = new THREE.PerspectiveCamera(
      50,
      window.innerWidth / window.innerHeight,
      0.1,
      20
    );
    camera.position.set(0, 2, 3);
    camera.lookAt(0, 2, -5);

    //Controls
    controls = new OrbitControls(camera, renderer.domElement);
    controls.target.set(0, 1, 0);
    controls.update();

    // lights
    const dirLight = new THREE.DirectionalLight('white',0.5);
    dirLight.position.set(0, 1, -0.4);
    dirLight.castShadow = false;
    scene.add(dirLight);

    const lightFromBack = new THREE.DirectionalLight('white',0.5);
    lightFromBack.position.set(0, 1, 0.8);
    lightFromBack.castShadow = true;
    scene.add(lightFromBack);

    const ambLight = new THREE.AmbientLight('white',2);
    scene.add(ambLight);

    // Geometry
    const sphere = new THREE.SphereGeometry(0.1);
    const targetMaterial = new THREE.MeshBasicMaterial({ color: 0xff0000 });
    targetGeometry = new THREE.Mesh(sphere, targetMaterial);
    targetGeometry.visible = false;
    scene.add(targetGeometry);

    var timeoutId;
    var pickDelay = 250;
    var isClick = true;
    container.onmousedown = function (event) {
      timeoutId = setTimeout(function () {
        isClick = false;
      }, pickDelay);
    };
    container.onmouseup = function (event) {
      clearTimeout(timeoutId);
      if (/*selectingCenter &&*/ isClick) {
        const rect = renderer.domElement.getBoundingClientRect();
        const x = event.clientX - rect.left;
        const y = event.clientY - rect.top;

        var mouse = new THREE.Vector2();
        var canvas = document.querySelector(".canvas");
        mouse.x = (x / canvas.clientWidth) * 2 - 1;
        mouse.y = (y / canvas.clientHeight) * - 2 + 1
        selectPickedBone(mouse);
      }
      isClick = true;
    }
    
    loadScene(scene);
  }

  function resize() {
    camera.aspect = window.innerWidth / window.innerHeight;
    camera.updateProjectionMatrix();
    renderer.setSize(window.innerWidth, window.innerHeight);
  }

  function selectPickedBone(mousePos) {
    camera.updateMatrixWorld();
    var raycaster = new THREE.Raycaster();
    raycaster.setFromCamera(mousePos, camera);

    var intersects = raycaster.intersectObjects(model.children, true);

    if (intersects.length > 0) {
      // Find nearest bone
      const pickedPoint = intersects[0].point;
      var minDist = Number.POSITIVE_INFINITY;
      findNearestBone(bones, pickedPoint, minDist);

      // visualize selection sphere
      targetGeometry.visible = true;
      centreBone.getWorldPosition(bonePos);
      targetGeometry.position.copy(bonePos);
      var selectDelay = 500;
      setTimeout(function () { targetGeometry.visible = false; }, selectDelay);
    }
  }

  function findNearestBone(bone, point, minDist) {
    var bonePosition = new THREE.Vector3();
    bone.getWorldPosition(bonePosition);
    var distance = point.distanceTo(bonePosition);
    if (distance < minDist) {
      minDist = distance;
      centreBone = bone;
    }

    for (var childBone of bone.children) {
      minDist = findNearestBone(childBone, point, minDist);
    }

    return minDist;
  }

  //Recursive animate-function
  function animate() {
    requestAnimationFrame(animate);

    let delta = clock.getDelta();
    if (singleStepMode) {
      delta = stepSize;
      stepSize = 0;
    }
    if(mixer)
      mixer.update(delta);

    if (sceneBB) {
      if (camera.position.x > sceneBB.max.x)
        camera.position.x = sceneBB.max.x;
      if (camera.position.x < sceneBB.min.x)
        camera.position.x = sceneBB.min.x;
      if (camera.position.y > sceneBB.max.y + 5)
        camera.position.y = sceneBB.max.y + 5;
      if (camera.position.y < sceneBB.min.y + 3)
        camera.position.y = sceneBB.min.y + 3;
      if (camera.position.z > sceneBB.max.z)
        camera.position.z = sceneBB.max.z;
      if (camera.position.z < sceneBB.min.z)
        camera.position.z = sceneBB.min.z;
    }

    if (centreBone) {
      centreBone.getWorldPosition(bonePos);
      controls.target.copy(bonePos);
    }

    renderer.render(scene, camera);
  }

  return (
    <>
      <ControlPanel />

      <div className="canvasParent">
        <canvas className="canvas"></canvas>
        <div className="optionDiv"></div>
        <div className="loadingDiv">
          <h1>Laster...</h1>
          <img
            height="300px"
            width="300px"
            src="/img/mesh_models.png"
            alt="dancingImage"
          />
        </div>
      </div>
      <audio loop>
        <source src="/mp3/Art-Of-Silence.mp3" type="audio/mpeg" />
      </audio>
    </>
  );
};

export default AnimationScene;
