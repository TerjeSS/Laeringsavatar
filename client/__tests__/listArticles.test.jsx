import { ListArticles } from "../pages/listArticles";

import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { ArticlesApiContext } from "../articlesApiContext";

describe("ListArticles component", () => {
  it("shows loading screen", () => {
    const domElement = document.createElement("div");
    ReactDOM.render(<ListArticles />, domElement);
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("shows articles", async () => {
    const articles = [{ topic: "topic 1" }, { topic: "topic 2" }];
    const domElement = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <ArticlesApiContext.Provider value={{ listArticles: () => articles }}>
          <ListArticles />
        </ArticlesApiContext.Provider>,
        domElement
      );
    });

    expect(
      Array.from(domElement.querySelectorAll("h3")).map((e) => e.innerHTML)
    ).toEqual(["topic 1", "topic 2"]);
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("shows error message", async () => {
    const domElement = document.createElement("div");
    await act(async () => {
      const listArticles = () => {
        throw new Error("Something went wrong");
      };
      ReactDOM.render(
        <ArticlesApiContext.Provider value={{ listArticles }}>
          <ListArticles />
        </ArticlesApiContext.Provider>,
        domElement
      );
    });

    expect(domElement.querySelector("#error-text").innerHTML).toEqual(
      "Error: Something went wrong"
    );
    expect(domElement.innerHTML).toMatchSnapshot();
  });
});
