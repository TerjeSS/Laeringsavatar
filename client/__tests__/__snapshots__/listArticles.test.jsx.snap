// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListArticles component shows articles 1`] = `"<div><div id=\\"article_card_info\\"><h1></h1><h3>topic 1</h3><p>Publisert: , </p><img alt=\\"Article poster\\" width=\\"300\\" height=\\"200\\"><p></p></div><div id=\\"article_card_info\\"><h1></h1><h3>topic 2</h3><p>Publisert: , </p><img alt=\\"Article poster\\" width=\\"300\\" height=\\"200\\"><p></p></div></div>"`;

exports[`ListArticles component shows error message 1`] = `"<div><h1>Error</h1><div id=\\"error-text\\">Error: Something went wrong</div></div>"`;

exports[`ListArticles component shows loading screen 1`] = `"<div>Loading...</div>"`;
